docs/develop/typescript/message-passing.mdx

```diff
@@ -196,107 +290,319 @@ await client.workflow.signalWithStart(yourWorkflow, {
+It doesn't apply to pure read operations, like Queries or Update Validators.
+
+:::tip
+
+For additional information, see [Inject work into the main Workflow](/encyclopedia/workflow-message-passing#injecting-work-into-main-workflow), [Ensuring your messages are processed exactly once](/encyclopedia/workflow-message-passing#exactly-once-message-processing), and [this sample](https://github.com/temporalio/samples-typescript/blob/main/pdates-and-signals/safe-message-handlers/README.md) demonstrating safe `async` message handling.
+
+:::
+
+### Use async handlers {#async-handlers}
+
+Signal and Update handlers can be `async` functions.
+Using `async` allows you to use `await` with Activities, Child Workflows, durable [`workflow.sleep`](https://typescript.temporal.io/api/namespaces/workflow#sleep) Timers, [`workflow.condition`](https://typescript.temporal.io/api/namespaces/workflow#condition) conditions, etc.
+This expands the possibilities for what can be done by a handler but it also means that handler executions and your main Workflow method are all running concurrently, with switching occurring between them at `await` calls.
+It's essential to understand the things that could go wrong in order to use `async` handlers safely.
+See [Workflow message passing](/encyclopedia/workflow-message-passing) for guidance on safe usage of async Signal and Update handlers, the [Safe message handlers](https://github.com/temporalio/samples-typescript/blob/main/pdates-and-signals/safe-message-handlers/README.md) sample, and the [Controlling handler concurrency](#control-handler-concurrency) and [Waiting for message handlers to finish](#wait-for-message-handlers) sections below.
+
+The following code executes an Activity that makes a network call to a remote service.
+It modifies the Update handler from earlier on this page, turning it into an `async def`:
+
+```typescript
+// 👉 Use the objects returned by defineUpdate to set the Update handler in
+// Workflow code, and to send Updates from Client code.
+export const setLanguageUsingActivity = wf.defineUpdate<Language, [Language]>('setLanguageUsingActivity');
+
+export async function greetingWorkflow(): Promise<string> {
+  const greetings: Partial<Record<Language, string>> = {
+    [Language.CHINESE]: '你好，世界',
+    [Language.ENGLISH]: 'Hello, world',
+  };
+
+  let language = Language.ENGLISH;
+
+  const lock = new Mutex();
+  wf.setHandler(setLanguageUsingActivity, async (newLanguage) => {
+    // 👉 An Update handler can mutate the Workflow state and return a value.
+    // 👉 Since this update handler is async, it can execute an activity.
+    if (!(newLanguage in greetings)) {
+      // 👉 Do the following with the lock held to ensure that multiple calls to set_language are processed in order.
+      await lock.runExclusive(async () => {
+        // Simulate an async call to a greeting service
+        await new Promise((resolve) => setTimeout(resolve, 200));
```

Comment:
```suggestion
        await wf.sleep(200)
```

Reply:
Thanks, that line shouldn't be there at all, in either form. Fixed.